src/server/folding_range.rs

```diff
@@ -0,0 +1,36 @@
+use tower_lsp::lsp_types::{FoldingRange, FoldingRangeKind};
+use typst::syntax::Source;
+
+use super::TypstServer;
+
+impl TypstServer {
+    pub fn get_folding_ranges(&self, source: &Source) -> Option<Vec<FoldingRange>> {
+        let mut ranges = vec![];
+        let mut previous_idx: Option<usize> = None;
+
+        for (line_idx, line) in source.text().lines().enumerate() {
+            if line.starts_with('=') {
```

Comment:
Can we get heading information using the AST? Detecting `==` is OK but fragile. The outline function also need to get all headings. This is how it is implemented: https://github.com/nvarner/typst-lsp/blob/master/src/server/symbols.rs#L98

Reply:
We can and imo we should (i.e. raw text lines can start with `=`

Reply:
Ah, that's a good point, I hadn't considered it. I'll try my hand at using the AST's information later this evening ­— as much as I'd prefer not to, I've got exams to study for as well :sweat_smile:

Reply:
Okay, that was a lot easier than I thought it'd be. Should be all good now :)